### DISCUSSION TITLE
next/link: How to prefetch page data (.json)?

### DISCUSSION THREAD

**Discussion by neongreen** [vercel/next.js#33131] [Help] [answered]
I want `<Link>` to prefetch page data (`_next/data/...`) in addition to the page .js itself. I don't mind the increased server load and would happily trade it off for page speed.

Is there any way to accomplish this?

(NB: there used to be an issue in Next for this, but then it got implemented as https://github.com/scaleapi/data-prefetch-link and the issue was closed. Now the data-prefetch-link package is outdated and I guess we're back to square 1 unless I missed something.)

**Update 1:** I tried rewriting my pages to use `getInitialProps` and then going the https://github.com/scaleapi/data-prefetch-link route, but `router.prefetch` no long returns the prefetched page .js in the promise and I can't call its `getInitialProps`.

**Update 2:** I think I'll just annotate each `<Link>` with the component corresponding to that link so that I can call `getInitialProps` on prefetch. We'll see how it goes. I still need to copy `next/link` into my project but it's not that big.

**Comment by sethsnel** [comment]
Funny thing, my site loads the .json data files (static props) for ALL my pages generated with `getStaticPaths`. And I'm scared it might get to large. So I actually want to prevent the app from pre-fetching all content. How do I get what you have? Because yours don't seem to load.

My site: -removed-
If you open the network tab you'll see all json files for all pages are downloaded.

**Reply by neongreen** [reply]
I'm using `getServerSideProps`, not `getStaticPaths`.

I would assume that those .json files are only loaded after the page renders (although you might want to check) so loading all those files shouldn't impact page speed even if they are large. The machinery in next/link also makes sure to not do preloading if the connection is slow. Or are you worried about something other than page speed?

**Reply by sethsnel** [reply]
I'm worried that in the future more pages are added and that those json contents also grow. And I don't want users to download the complete repository, let's say if it grows to 25mb of json files.

**Reply by neongreen** [reply]
next/link is supposed to preload only links in the viewport, not all pages anywhere on the site.

You can also use `<Link prefetch={false}>` to disable prefetching on page load (it would still happen on link hover though).

**Reply by sethsnel** [reply]
We are developing a wiki-kind of app. So there could be many links in the viewport which we don't want to initially load. Loading on hover is fine though.

**Comment by neongreen** [comment, marked as answer]
I have migrated all pages to use `getInitialProps` and [react-query](https://react-query.tanstack.com/). I also have a custom `<LinkPreload>` that is based on `<Link>` and calls the `.preload` method of each page it tries to preload.

**Reply by camjw** [reply]
Hey @neongreen - do you think you'd be able to share your LinkPreload? Sounds super helpful.

**Comment by neongreen** [comment]
Yeah but it’s not a great option b/c it’s a bunch of code copied from a specific version of Next and modified. If you don’t upgrade Next.js then it’s fine, if you do then you probably have to backport their changes.

Also admittedly it doesn’t work as well as it should. Creates a preload on any link hover even if the previous preload happened a millisecond ago.

This said, here you go: https://gist.github.com/neongreen/9c888f98110842feede1bc2dc2d3d522 <https://gist.github.com/neongreen/9c888f98110842feede1bc2dc2d3d522>

My modifications are under whatever license Next itself is, use them however you wish.

**Reply by camjw** [reply]
Thank you so much!
